### PR TITLE
Surface mesh approximation: handle degenerate cases

### PR DESCRIPTION
**Pull request by aboudev**
## Fix crash due to degenerate cases

Handle degenerate faces, degenerate 1-d boundaries and floating ball-components with one proxy.

## Release Management

* Affected package(s): Surface Mesh Approximation

### CODE REVIEW

File: Polyhedron/demo/Polyhedron/Plugins/Surface_mesh/VSA_wrapper.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@ VSA_wrapper::VSA_wrapper(const SMesh &mesh) :
     const Point_3 &p2 = vpm[target(next(he, mesh), mesh)];
 
     put(m_center_pmap, f, CGAL::centroid(p0, p1, p2));
-    put(m_area_pmap, f, std::sqrt(CGAL::to_double(CGAL::squared_area(p0, p1, p2))));
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Plugins/Surface_mesh/VSA_wrapper.cpp]: I'm surprised you need approximate_sqrt in the demo as the kernel is EPICK

File: Surface_mesh_approximation/examples/Surface_mesh_approximation/vsa_isotropic_metric_example.cpp [hunk 2]
```diff
@@ -85,7 +84,7 @@ int main()
     const Point_3 &p0 = vpmap[source(he, mesh)];
     const Point_3 &p1 = vpmap[target(he, mesh)];
     const Point_3 &p2 = vpmap[target(next(he, mesh), mesh)];
-    put(area_pmap, f, FT(std::sqrt(CGAL::to_double(CGAL::squared_area(p0, p1, p2)))));
+    put(area_pmap, f, CGAL::approximate_sqrt(CGAL::squared_area(p0, p1, p2)));
```
Comment (sloriot) [inline Surface_mesh_approximation/examples/Surface_mesh_approximation/vsa_isotropic_metric_example.cpp]: same here, kernel is EPICK

File: Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h [hunk 3]
```diff
@@ -1429,13 +1424,17 @@ class Variational_shape_approximation {
         const Point_3 &p0 = m_vpoint_map[source(he, *m_ptm)];
         const Point_3 &p1 = m_vpoint_map[target(he, *m_ptm)];
         const Point_3 &p2 = m_vpoint_map[target(next(he, *m_ptm), *m_ptm)];
-        const FT farea(std::sqrt(CGAL::to_double(CGAL::squared_area(p0, p1, p2))));
-        const Vector_3 fnorm = CGAL::unit_normal(p0, p1, p2);
+        const FT farea = CGAL::approximate_sqrt(CGAL::squared_area(p0, p1, p2));
+        const Vector_3 fnorm =
+          CGAL::collinear(p0, p1, p2) ? CGAL::NULL_VECTOR : CGAL::unit_normal(p0, p1, p2);
```
Comment (sloriot) [inline Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h]: Note that you should use the function Collinear_3 like you are doing with other functors (l141 for example).

File: Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h [hunk 4]
```diff
@@ -1545,20 +1544,35 @@ class Variational_shape_approximation {
...
-        vec = CGAL::cross_product(chord_vec, vec);
-        FT dist(std::sqrt(CGAL::to_double(vec.squared_length())));
-        if (dist > dist_max) {
-          dist_max = dist;
-          he_max = he;
+      if (chord_vec.squared_length() > FT(0.0)) {
+        FT dist_max(0.0);
+        chord_vec = scale_functor(chord_vec,
+          FT(1.0) / CGAL::approximate_sqrt(chord_vec.squared_length()));
+        BOOST_FOREACH(const halfedge_descriptor &he, chord) {
```
Comment (sloriot) [inline Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h]: take by copy not ref

File: Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h [hunk 5]
```diff
@@ -1545,20 +1544,35 @@ class Variational_shape_approximation {
...
+          const FT dist = CGAL::approximate_sqrt(vec.squared_length());
+          if (dist > dist_max) {
+            dist_max = dist;
+            he_max = he;
+          }
+        }
+      }
+      else {
+        FT dist_max(0.0);
+        BOOST_FOREACH(const halfedge_descriptor &he, chord) {
```
Comment (sloriot) [inline Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h]: same here

File: Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h [hunk 6]
```diff
@@ -1800,21 +1828,34 @@ class Variational_shape_approximation {
...
-        if (dist > dist_max) {
-          chord_max = citr;
-          dist_max = dist;
+      const FT chord_len = CGAL::approximate_sqrt(chord_vec.squared_length());
+      bool degenerate_chord = false;
+      if (chord_len > FT(0.0)) {
+        chord_vec = scale_functor(chord_vec, FT(1.0) / chord_len);
+        for (Boundary_chord_iterator citr = chord_begin; citr != chord_end; ++citr) {
+          Vector_3 vec = vector_functor(pt_begin, m_vpoint_map[target(*citr, *m_ptm)]);
+          vec = CGAL::cross_product(chord_vec, vec);
```
Comment (sloriot) [inline Surface_mesh_approximation/include/CGAL/Variational_shape_approximation.h]: Use the functor?

**Comment by aboudev** [comment]
Thanks for @sloriot 's review. New commits should solve the issues.

**Comment by maxGimeno** [comment]
Successfully tested by https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-136.shtml

**Pull request merged**